### PR TITLE
TW-109: add fetch contacts screen when create new group or private chat

### PR DESCRIPTION
**Pull request by sherlockvn**
https://github.com/linagora/twake-on-matrix/assets/43041967/c27f1dcd-3f5b-4ad7-af4d-a636336736c0

https://github.com/linagora/twake-on-matrix/assets/43041967/fa583368-3a9d-45e8-ae45-c52e79f46d55

### CODE REVIEW

File: lib/pages/new_group/new_group_chat_info.dart [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: _buildAppBar(context),
+      body: Padding(
+        padding: const EdgeInsets.symmetric(vertical: 16.0, horizontal: 8.0),
+        child: Column(
+          children: [
+            _buildChangeProfileWidget(context),
+            const SizedBox(height: 16.0,),
```
Comment (imGok) [inline lib/pages/new_group/new_group_chat_info.dart:34]: Could you fix this indentation

File: lib/pages/new_group/new_group_chat_info.dart [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            Text(L10n.of(context)!.addAPhoto,
+              style: Theme.of(context).textTheme.labelLarge?.copyWith(
+                color: Theme.of(context).colorScheme.onSurface,
+              ),
+            ),
+            Text(L10n.of(context)!.maxImageSize(5),
+              style: Theme.of(context).textTheme.bodySmall?.copyWith(
+                color: LinagoraRefColors.material().neutral[40],
+              ),),
+            const SizedBox(height: 32,),
```
Comment (imGok) [inline lib/pages/new_group/new_group_chat_info.dart:44]: same

File: lib/pages/new_group/new_group_chat_info.dart [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            _buildGroupNameTextFieid(context),
+            const SizedBox(height: 16,),
```
Comment (imGok) [inline lib/pages/new_group/new_group_chat_info.dart:46]: same

File: lib/pages/new_group/new_group_chat_info.dart [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+          ],
+        ),
+      ),
+      floatingActionButton: ValueListenableBuilder<bool>(
+        valueListenable: newGroupController.haveGroupNameNotifier, 
+        builder: (context, value, child) {
+          if (!value) {
+            return const SizedBox.shrink();
+          }
+          return child!;
```
Comment (imGok) [inline lib/pages/new_group/new_group_chat_info.dart:60]: Why do you need to force non-nullable?

File: lib/pages/new_group/new_group_chat_info.dart [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        ),
+      ),
+    );
+  }
+
+  PreferredSizeWidget _buildAppBar(BuildContext context) {
+    return PreferredSize(
+      preferredSize: const Size.fromHeight(64.0),
+      child: AppBar(
+        toolbarHeight: 64,
```
Comment (imGok) [inline lib/pages/new_group/new_group_chat_info.dart:74]: Can you make a style file with important styles not all of them ?

File: lib/pages/new_group/widget/expansion_participants_list.dart [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  State<StatefulWidget> createState() => _ExpansionParticipantsListState();
+}
+
+class _ExpansionParticipantsListState extends State<ExpansionParticipantsList> {
+  bool isExpanded = true;
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.only(left: 0.0, right: 0.0),
```
Comment (imGok) [inline lib/pages/new_group/widget/expansion_participants_list.dart:27]: fyi, you can do `EdgeInsets.symmetric(horizontal: 0)` if you find it cleaner

File: lib/pages/new_group/widget/expansion_participants_list.dart [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                    child: Row(
+                      children: [
+                        Text(L10n.of(context)!.participantsCount(widget.contactsList.length),
+                          style: Theme.of(context).textTheme.labelLarge?.copyWith(
+                            color: LinagoraRefColors.material().neutral[40],
+                          ),),
+                      ],
+                    ),
+                  ),
+                  const Expanded(child: SizedBox.shrink()),
```
Comment (imGok) [inline lib/pages/new_group/widget/expansion_participants_list.dart:48]: Isn't it the same utility as "Spacer()" ?

File: lib/pages/new_private_chat/widget/loading_contact_widget.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  const LoadingContactWidget({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Center(
+      child: Column(
+        children: [
+          const CircularProgressIndicator(),
+          const SizedBox(height: 16.0,),
+          Text("Loading contacts...",
```
Comment (imGok) [inline lib/pages/new_private_chat/widget/loading_contact_widget.dart:13]: We don't have international string for this ? :(

File: lib/pages/new_group/new_group_chat_info.dart (hunk 2 above)
Comment (sherlockvn) [reply lib/pages/new_group/new_group_chat_info.dart:44]: IMO, you should review in IDE, in my side, i see no problem with indentation

File: lib/pages/new_private_chat/widget/loading_contact_widget.dart (hunk 8 above)
Comment (sherlockvn) [reply lib/pages/new_private_chat/widget/loading_contact_widget.dart:13]: sr :))

File: lib/pages/new_group/widget/expansion_participants_list.dart (hunk 6 above)
Comment (sherlockvn) [reply lib/pages/new_group/widget/expansion_participants_list.dart:27]: okey

File: lib/pages/new_group/widget/expansion_participants_list.dart (hunk 7 above)
Comment (sherlockvn) [reply lib/pages/new_group/widget/expansion_participants_list.dart:48]: no. it different.

File: lib/widgets/setting_tile.dart [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                  if (settingDescription != null && settingDescription!.isNotEmpty)
+                    Text(settingDescription!,
+                      style: Theme.of(context).textTheme.bodySmall?.copyWith(
+                        letterSpacing: 0.4,
+                    ),),
+                ],
+              ),
+            ),
+            const SizedBox(width: 12,),
+            IgnorePointer(
```
[the remark this answers is not in the document]
Comment (sherlockvn) [reply lib/widgets/setting_tile.dart:67]: i use it to allow enable/disable the switch button in setting_tile. No i don't think it bad for testing in the future.

**Comment by hoangdat** [comment]
I see it too height, it decrease the main screen a lot

Screenshot 2023-06-04 at 16 05 57

File: assets/l10n/intl_en.arb [hunk 10]
```diff
@@ -2618,5 +2618,10 @@
             "count": {}
         }
     },
-    "back": "Back"
+    "back": "Back",
+    "contacts": "Contacts",
+    "searchForContacts": "Search for contacts",
+    "soonThereHaveContacts": "Soon there will be contacts",
```
Comment (hoangdat) [inline assets/l10n/intl_en.arb:2624]: we use this for what?

File: lib/config/app_config.dart [hunk 11]
```diff
@@ -41,7 +41,7 @@ abstract class AppConfig {
   static bool hideUnknownEvents = true;
   static bool hideUnimportantStateEvents = true;
   static bool showDirectChatsInSpaces = true;
-  static bool separateChatTypes = false;
+  static bool separateChatTypes = true;
```
Comment (hoangdat) [inline lib/config/app_config.dart:44]: what use for?

File: lib/data/model/presentation_contact.dart
Comment (hoangdat) [inline lib/data/model/presentation_contact.dart:1]: Why we move this presentation model to `data` layer?

File: lib/pages/chat_list/chat_list.dart [hunk 12]
```diff
@@ -47,7 +56,32 @@ enum ActiveFilter {
...
   groups,
   messages,
-  spaces,
+  spaces
+}
+
+enum BottomTabbar {
+  chats(tabIndex: 1),
+  contacts(tabIndex: 0),
+  stories(tabIndex: 2);
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list.dart:65]: why not a separated VRoute Element for each tab?

File: lib/data/model/presentation_contact.dart
Comment (hoangdat) [reply lib/data/model/presentation_contact.dart:1]: `lib/preseantation/model`

**Comment by sherlockvn** [comment]
> I see it too height, it decreases the main screen a lot
> 
> Screenshot 2023-06-04 at 16 05 57

Resolved

Screenshot 2023-06-05 at 13 19 00
Screenshot 2023-06-05 at 13 20 04

File: lib/config/routes.dart [hunk 13]
```diff
@@ -46,77 +48,117 @@ class AppRoutes {
...
+              VWidget(
+                path: '/stories', 
+                widget: const Scaffold(),
+                buildTransition: _bottomToTopTransition,
+              ),
+            ],
+          ),
+          VPopHandler(
+            onPop: (vRedirector) async {
+              vRedirector.to('/rooms');
```
Comment (hoangdat) [inline lib/config/routes.dart:78]: you mean: whenever you `pop` in the widget inside, chat list will be open again?

File: lib/pages/new_private_chat/widget/loading_contact_widget.dart [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+class LoadingContactWidget extends StatelessWidget {
+  const LoadingContactWidget({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return Center(
+      child: Column(
+        children: [
+          const CircularProgressIndicator(),
```
Comment (hoangdat) [inline lib/pages/new_private_chat/widget/loading_contact_widget.dart:12]: please add more gap from the appbar by adding above:
```
const SizedBox(height: 16.0,),
```

File: lib/pages/new_private_chat/search_contacts_controller.dart [hunk 15]
```diff
@@ -56,8 +62,15 @@ class SearchContactsController {
...
   }
 
+  void onCloseSearchTapped() {
+    textEditingController.clear();
+  }
+
   void dispose() {
     _debouncer.cancel();
+    textEditingController.removeListener(() {
+      Logs().d("SearchContactsController: dispose(): remove Listerners");
```
Comment (hoangdat) [inline lib/pages/new_private_chat/search_contacts_controller.dart:72]: why we need it?

File: lib/pages/new_private_chat/widget/loading_contact_widget.dart (hunk 14 above)
Comment (sherlockvn) [reply lib/pages/new_private_chat/widget/loading_contact_widget.dart:13]: Done!

File: assets/l10n/intl_en.arb [hunk 16]
```diff
@@ -2621,5 +2621,10 @@
     "back": "Back",
     "loginWithSSO": "Log in with SSO",
     "wrongServerName": "Wrong server name",
-    "serverNameWrongExplain": "Server address was sent to you by company admin. Check the invitation email."
+    "serverNameWrongExplain": "Server address was sent to you by company admin. Check the invitation email.",
+    "contacts": "Contacts",
+    "searchForContacts": "Search for contacts",
+    "soonThereHaveContacts": "Soon there will be contacts",
+    "searchSuggestion": "For now, search by typing a person’s name or public server address",
+    "loadingContacts": "Loading contacts..."
```
Comment (imGok) [inline assets/l10n/intl_en.arb:2629]: You should use "…" special character to be consistent with other strings

**Review by hoangdat — APPROVED** [review]

File: lib/pages/chat_list/chat_list_view.dart [hunk 17]
```diff
@@ -1,212 +1,43 @@
...
+              VRouter.of(context).to('/newprivatechat'),
+          helpLabel: L10n.of(context)!.newChat,
+          child: TwakeFloatingActionButton(
+            icon: Icons.mode_edit_outline_outlined,
+            onTap: () => VRouter.of(context).to('/newprivatechat'),
           ),
-        );
-      },
+        )
+        : null,
```
Comment (imGok) [inline lib/pages/chat_list/chat_list_view.dart:40]: Could you replace this by `const SizedBox.shrink()`

File: lib/pages/chat_list/home_screen.dart [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+class _HomeScreenState extends State<HomeScreen> {
+  int selectedIndex = 1;
+
+  void onDestinationSelected(int selected) {
+    setState(() {
+      selectedIndex = selected;
+      VRouter.of(context).to(BottomTabbar.fromIndex(selected).path);
+      if (selected == BottomTabbar.stories.tabIndex) {
+        Fluttertoast.showToast(
+          msg: "Stories is not ready yet!",
```
Comment (imGok) [inline lib/pages/chat_list/home_screen.dart:29]: Why are we not using an international string for this ?

File: lib/pages/new_private_chat/search_contacts_controller.dart [hunk 19]
```diff
@@ -18,12 +18,18 @@ class SearchContactsController {
...
   String searchKeyword = "";
 
   void init() {
     _initializeDebouncer();
     textEditingController.addListener(() {
+      if (textEditingController.text.isNotEmpty) {
+        isSearchModeNotifier.value = true;
+      } else {
+        isSearchModeNotifier.value = false;
+      }
```
Comment (imGok) [inline lib/pages/new_private_chat/search_contacts_controller.dart:32]: `isSearchModeNotifier.value = textEditingController.text.isNotEmpty`  ?

File: lib/pages/chat_list/home_screen.dart (hunk 18 above)
Comment (hoangdat) [reply lib/pages/chat_list/home_screen.dart:29]: IMO, it is temporary, no need to use i10n

**Pull request merged**